### ISSUE TITLE
[Bug?]: VS Code on macOS unable to resolve typescript imports - "Cannot find module"

### ISSUE THREAD

**Issue by joshhunt** [yarnpkg/berry#5814]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

When trying to upgrade to Typescript 5.2.2 (from 4.8.4), VS Code is unable to resolve types for imports when running on macOS and shows the error message "Cannot find module '@emotion/css' or its corresponding type declarations.ts(2307)". On Windows (WSL) or Linux, it works fine. Running `yarn tsc` to check the types works as expected.

Using yarn@3.6.4 with the pnp linker.

<img width="1319" alt="Screenshot 2023-10-20 at 10 43 04 am" src="https://github.com/yarnpkg/berry/assets/46142/3bef182b-8215-4186-bf22-1488e46e5251">

### To reproduce

Currently I've been unable to reproduce this on an extremely minimal reproduction repo. The issue only seems to occur in our repository.

To reproduce, clone grafana and check out the `academo/typescript-5.x` branch on macOS:

1. Clone the [grafana](https://github.com/grafana/grafana) repo with `git clone git@github.com:grafana/grafana.git`, or whatever your preferred method is
2. `cd grafana`
3. `yarn install`
4. Check out the [typescript 5 branch](https://github.com/grafana/grafana/pull/76511) with `git checkout academo/typescript-5.x` or `gh pr checkout 76511`
5. `yarn install`
6. Open VS Code, ensure that the editor is using the workspace 5.2.2-sdk version of typescript
7. Open [`public/app/features/browse-dashboards/BrowseDashboardsPage.tsx`](https://github.com/grafana/grafana/blob/e84599bd51706138c40b61582c8d9e5b5b3ad7b1/public/app/features/browse-dashboards/BrowseDashboardsPage.tsx) and observe the red squiggles 

If there's something suspected that might cause the issue that we could put into a more minimal reproduction, I'm happy to try and put that together.

### Environment

```shell
System:
    OS: macOS 13.5.2
    CPU: (10) arm64 Apple M1 Pro
  Binaries:
    Node: 18.12.0 - /private/var/folders/ch/51232r6s0pjdzy8rmzhpmk7h0000gn/T/xfs-197180fa/node
    Yarn: 3.6.4 - /private/var/folders/ch/51232r6s0pjdzy8rmzhpmk7h0000gn/T/xfs-197180fa/yarn
    npm: 8.19.2 - ~/Library/Caches/fnm_multishells/23592_1697193197983/bin/npm
```

### Additional context

We're reproduced this breaking for two different people/machines running macOS, and working correctly with one person on Linux, and another person running the repo/vs code remotely on Windows using WSL.

**Comment by currybob** [comment]
Is there any updates for this issue?

**Comment by joshhunt** [comment]
We were facing this issue with Yarn 3.6.4. We updated to Yarn 4 and the issue no longer occurred.

It's disappointing to have opaque problems like this where the package manager gets in the way.

**Comment by danielrsantana** [comment]
I'm having this issue with the 4.0.2, with all external references. It's a monorepo, NodeJS Express Typescript (server) and a React App (web-app).

How I did the Migration:

- `corepack enable`
- `yarn set version berry`
-  reviewed `.yarnrc.yml`
- `yarn dlx @yarnpkg/sdks vscode` (in the root of my monorepo)
- ctrl+shift+p in a TypeScript file, "Select TypeScript Version", "Use Workspace Version"

<img width="739" alt="image" src="https://github.com/yarnpkg/berry/assets/11296189/007efb66-111f-4d80-bf19-d570c553323f">

<img width="884" alt="image" src="https://github.com/yarnpkg/berry/assets/11296189/d9fc2c8c-f878-4dc8-99e3-1a1371c05d08">

Just in time: my package.json:

```
{
  "name": "some-server",
  "version": "v1.6.00beta",
  "main": "dist/index.js",
  "license": "UNLICENSED",
  "private": true,
  "scripts": {
    "start": "nodemon src/index.ts",
    "build": "tsc",
  },
  "dependencies": {
    "body-parser": "^1.20.2",
    "cors": "^2.8.5",
    "dotenv": "^16.3.1",
    "express": "^4.18.2",
    "express-jwt": "^8.4.1",
    "helmet": "^7.1.0",
    "jsonwebtoken": "^9.0.2",
    "jwks-rsa": "^3.1.0",
    "mongodb": "^6.3.0",
    "multer": "^1.4.5-lts.1",
    "openai": "^4.21.0"
  },
  "devDependencies": {
    "@types/body-parser": "^1.19.5",
    "@types/cors": "^2.8.17",
    "@types/express": "^4.17.21",
    "@types/multer": "^1.4.11",
    "@types/node": "^20.10.4",
    "@typescript-eslint/eslint-plugin": "^6.14.0",
    "@typescript-eslint/parser": "^6.14.0",
    "eslint": "^8.55.0",
    "eslint-config-airbnb-typescript": "^17.1.0",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-import": "^2.29.1",
    "eslint-plugin-prettier": "^5.0.1",
    "jest": "^29.7.0",
    "nodemon": "^3.0.2",
    "prettier": "^3.1.1",
    "ts-node": "^10.9.2",
    "typescript": "^5.3.3"
  },
  "packageManager": "yarn@4.0.2"
}

```

**Comment by torkelo** [comment]
getting the same issue with latest yarn 4.0.2 and Typescript 5.3.3

**Comment by Luke-lsm** [comment]
> getting the same issue with latest yarn 4.0.2 and Typescript 5.3.3

Hello, I managed to sort this out using yarn 4.0.2 & TS 5.33 by making sure that I had the [ZipFS](https://marketplace.visualstudio.com/items?itemName=arcanis.vscode-zipfs) extension

Then running at the root of a monorepo
`yarn dlx @yarnpkg/sdks vscode` 

and using the workspace version of TS

**Comment by kelvin2200** [comment]
same problem with yarn 4.1.0 and typescript 5.4.2 + yarn workspaces

**Comment by drikusroor** [comment]
> > getting the same issue with latest yarn 4.0.2 and Typescript 5.3.3
> 
> Hello, I managed to sort this out using yarn 4.0.2 & TS 5.33 by making sure that I had the [ZipFS](https://marketplace.visualstudio.com/items?itemName=arcanis.vscode-zipfs) extension
> 
> Then running at the root of a monorepo `yarn dlx @yarnpkg/sdks vscode`
> 
> and using the workspace version of TS

Additionally, if your project and Yarn PnP are located in a subdirectory, make sure to set the right path in your VS Code's `settings.json`:

```js
// .vscode/settings.json
{
  "typescript.tsdk": "{subdirectory}/.yarn/sdks/typescript/lib",
  "typescript.enablePromptUseWorkspaceTsdk": true
}
```

See also https://stackoverflow.com/a/67403641/4496102

**Comment by rkfg** [comment]
Happens on Linux as well with `typescript` 5.4, if the version is pinned to 5.3.3 (the last 5.3 version) it works. Tried `5.4.0-beta` (the first stable 5.4 version?) and up to 5.4.4, they're all broken. I couldn't find any other information about this bug, there are no distinct errors as well. Not only the imports are broken in the editor, `yarn tsc` also complains: `Cannot find module 'react' or its corresponding type declarations.` and so on. I'm unsure who's to blame: yarn, pnp, or typescript.

Can be trivially reproduced with:
```
yarn create vite testproject --template react-ts
cd testproject
yarn
yarn tsc
```
The version in `package.json` is set as `"typescript": "^5.2.2"` which causes 5.4.4 to be installed instead. Setting it to `5.3.3`, running `yarn` and then `yarn tsc` works fine.

**Comment by hunter-xue** [comment]
try to run: `yarn dlx @yarnpkg/sdks vscode`

**Comment by rkfg** [comment]
Yeah, this problem still stays after that. The SDK version matches the one in package.json and if it's from the 5.4 branch tsc/tsserver both break.

**Comment by rkfg** [comment]
Okay, I solved it. I have yarn 4.0.2 installed from Debian repo, there must be some bug in it. I did `yarn set version stable && yarn && yarn build` and it built the project just fine. `yarn dlx @yarnpkg/sdks vscode` also now installs a local TS version 5.4.4 that works well in VS Code.
